Revit_Core_Engine/Convert/Environment/ToRevit/Duct.cs

```diff
@@ -0,0 +1,210 @@
+
+                double diameter = elementProfile.Diameter;
+                revitDuct.SetParameter(BuiltInParameter.RBS_CURVE_DIAMETER_PARAM, diameter);
+
+                // Set LiningProfile
+                TubeProfile liningProfile = sectionProfile.LiningProfile as TubeProfile;
+                if (liningProfile != null)
+                {
+                    double liningThickness = liningProfile.Thickness;
+
+                    //Get first available ductLiningType from document
+                    Autodesk.Revit.DB.Mechanical.DuctLiningType dlt = new FilteredElementCollector(document).OfClass(typeof(Autodesk.Revit.DB.Mechanical.DuctLiningType)).FirstOrDefault() as Autodesk.Revit.DB.Mechanical.DuctLiningType;
+                    if (dlt == null)
+                    {
+                        BH.Engine.Reflection.Compute.RecordError("Any duct lining type needs to be present in the Revit model in order to push ducts with lining.");
+                        return null;
+                    }
+                    //Create ductLining
+                    Autodesk.Revit.DB.Mechanical.DuctLining dl = Autodesk.Revit.DB.Mechanical.DuctLining.Create(document, revitDuct.Id, dlt.Id, liningThickness);
+                }
+
+                // Set InsulationProfile
+                TubeProfile insulationProfile = sectionProfile.InsulationProfile as TubeProfile;
+                if (insulationProfile != null)
+                {
+                    double insulationThickness = insulationProfile.Thickness;
+
+                    // Get first available ductLiningType from document
+                    Autodesk.Revit.DB.Mechanical.DuctInsulationType dit = new FilteredElementCollector(document).OfClass(typeof(Autodesk.Revit.DB.Mechanical.DuctInsulationType)).FirstOrDefault() as Autodesk.Revit.DB.Mechanical.DuctInsulationType;
+                    if (dit == null)
+                    {
+                        BH.Engine.Reflection.Compute.RecordError("Any duct insulation type needs to be present in the Revit model in order to push ducts with insulation.");
+                        return null;
+                    }
+                    // Create ductInsulation
+                    Autodesk.Revit.DB.Mechanical.DuctInsulation di = Autodesk.Revit.DB.Mechanical.DuctInsulation.Create(document, revitDuct.Id, dit.Id, insulationThickness);
+                }
+                //Parameters not found in RevitAPI for Round Ducts, needs further investigation. Not critical for current workflows. 
+                //double profileThickness = elementProfile.Thickness;
+                //Oval Ducts are currently unsupported in the BHoM. No ShapeProfile exists. 
+            }
+            else
+            {
+                BH.Engine.Reflection.Compute.RecordError("No objects created. Only Box or TubeProfiles are supported.");
+                return null;
+            }
+
+            revitDuct = Duct.Create(document, mst.Id, ductType.Id, level.Id, start, end);
```

Comment:
As above, the object needs to be created before you start moving it or change its parameters.